File: app/src/main/java/in/example/rahul/vegcartpro/HomeActivity.java
package in.example.rahul.vegcartpro;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.daimajia.slider.library.Animations.DescriptionAnimation;
import com.daimajia.slider.library.SliderLayout;
import com.daimajia.slider.library.SliderTypes.BaseSliderView;
import com.daimajia.slider.library.SliderTypes.TextSliderView;
import com.daimajia.slider.library.Tricks.ViewPagerEx;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class HomeActivity extends AppCompatActivity implements BaseSliderView.OnSliderClickListener, ViewPagerEx.OnPageChangeListener, MyAdapter.ClickListener, NavigationView.OnNavigationItemSelectedListener {

    private List<Item> itemList= new ArrayList<>();
    private RecyclerView recyclerView;
    private MyAdapter adapter;
    private LinearLayout main;
    SliderLayout sliderLayout;
    HashMap<String, String> HashMapForURL;
    HashMap<String, Integer> HashMapForLocal;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);

        Toolbar toolbar=(Toolbar)findViewById(R.id.toolbar);
        toolbar.setTitle("Menu");
        sliderLayout=(SliderLayout)findViewById(R.id.slider);
        DrawerLayout drawer=(DrawerLayout) findViewById(R.id.drawer_layout);

       ActionBarDrawerToggle toggle=new ActionBarDrawerToggle(this,drawer,toolbar,R.string.open,R.string.close);

        drawer.setDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = (NavigationView)findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);
        View headerView=navigationView.getHeaderView(0);
       TextView txtFullName=(TextView)headerView.findViewById(R.id.txtFullName);
        // recyclerView=(RecyclerView)findViewById(R.id.recyclerView);
       // layoutManager.
        // Call below method if you want images online
        //AddImagesUrlOnline();

        // Call below method if you want images offline
       AddImageUrlLocal();

        // Call below method if you want to stop Automatic sliding
        //sliderLayout.stopAutoCycle();

        for (String name : HashMapForLocal.keySet()){
            TextSliderView textSliderView= new TextSliderView(HomeActivity.this);

            textSliderView.description(name).image(HashMapForLocal.get(name)).setScaleType(BaseSliderView.ScaleType.CenterCrop).setOnSliderClickListener(this);

            textSliderView.bundle(new Bundle());
            textSliderView.getBundle().putString("extra",name);
            sliderLayout.addSlider(textSliderView);
        }
        sliderLayout.setPresetTransformer(SliderLayout.Transformer.DepthPage);
        sliderLayout.setPresetIndicator(SliderLayout.PresetIndicators.Center_Bottom);
        sliderLayout.setCustomAnimation(new DescriptionAnimation());
        sliderLayout.setDuration(3000);
        sliderLayout.addOnPageChangeListener(HomeActivity.this);

        // Bottom
        recyclerView=(RecyclerView)findViewById(R.id.recyclerView);
        prepareItem();

        adapter= new MyAdapter(itemList);
        adapter.setClickListener(this);

        //RecyclerView.LayoutManager layoutManager= new LinearLayoutManager(getApplicationContext());
// Try Grid layout for Recycler view.
        RecyclerView.LayoutManager layoutManager=new GridLayoutManager(this,2,GridLayoutManager.VERTICAL,false);

        recyclerView.setLayoutManager(layoutManager);

        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setAdapter(adapter);
    }
    public void AddImagesUrlOnline(){
        HashMapForURL= new HashMap<String, String>();

        HashMapForURL.put("All Veg","http://");
    }
    public void AddImageUrlLocal(){
        HashMapForLocal= new HashMap<String, Integer>();

        HashMapForLocal.put("Fruits",R.drawable.fruit);
        HashMapForLocal.put("Cauliflower",R.drawable.cauliflower);
        HashMapForLocal.put("Fruit and Vegi",R.drawable.giphy);
        HashMapForLocal.put("Dancing Vegi",R.drawable.vegjig);


    }

    @Override
    protected void onStop() {
        sliderLayout.stopAutoCycle();
        super.onStop();
    }

    @Override
    public void onSliderClick(BaseSliderView slider) {
        Toast.makeText(this,slider.getBundle().get("extra")+ "",Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

    }

    @Override
    public void onPageSelected(int position) {

        Log.d("Slider Demo", "Page Change:" +position);
    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }
    public void foodVisit(View view){
        Intent intent= new Intent(HomeActivity.this, FoodActivity.class);
        startActivity(intent);
    }

    // Bottom

    private void prepareItem(){
        Item item= new Item(R.drawable.shutterstock,"Flower","Price is 30");
        itemList.add(item);

        item= new Item(R.drawable.apple,"Fruit","Price is 150");
        itemList.add(item);

        item= new Item(R.drawable.lettuce2,"Leaves","Price is 70");
        itemList.add(item);
        item= new Item(R.drawable.carrot,"Root","Price is 30");
        itemList.add(item);

        item= new Item(R.drawable.tomato,"Salad","Price is 20");
        itemList.add(item);

    }

    @Override
    public void itemClicked(View view, int position){
        if (position==0){
            Intent intent= new Intent(HomeActivity.this, FlowerVegActivity.class);
            intent.putExtra("ItemPosition ", position);
            startActivity(intent);
        }
        else if (position == 1){
            Intent intent= new Intent(HomeActivity.this, FruitActivity.class);
            intent.putExtra("ItemPosition ", position);
            startActivity(intent);
        }
        else if (position == 2){
            Intent intent= new Intent(HomeActivity.this, LeavesActivity.class);
            intent.putExtra("ItemPosition ", position);
            startActivity(intent);
        }
        else if (position == 3){
            Intent intent= new Intent(HomeActivity.this, RootActivity.class);
            intent.putExtra("ItemPosition ", position);
            startActivity(intent);
        }
        else if (position == 4){
            Intent intent= new Intent(HomeActivity.this, SaladActivity.class);
            intent.putExtra("ItemPosition ", position);
            startActivity(intent);
        }
        else {
            System.out.println("position..."+position);
        }
    }

    // Navigation Drawer
    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.home,menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        return super.onOptionsItemSelected(item);
    }
    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {

        int id = item.getItemId();

        if (id == R.id.fruit){
           Intent intent = new Intent(HomeActivity.this,FruitActivity.class);

           startActivity(intent);
        }
        else if (id == R.id.salid){
            Intent intent = new Intent(HomeActivity.this,SaladActivity.class);

            startActivity(intent);
        }
        else if (id == R.id.flower){
           Intent intent = new Intent(HomeActivity.this,FlowerVegActivity.class);

           startActivity(intent);
        }
        else if (id == R.id.leaves){
            Intent intent = new Intent(HomeActivity.this,LeavesActivity.class);

            startActivity(intent);
        }
        else if (id == R.id.root){
            Intent intent = new Intent(HomeActivity.this,RootActivity.class);

            startActivity(intent);
        }
        else if (id == R.id.nav_send){
            Intent intent = new Intent(HomeActivity.this,ContactActivity.class);

            startActivity(intent);
        }
        else if (id == R.id.mycart){
            Intent intent = new Intent(HomeActivity.this,MyCart.class);

            startActivity(intent);
        }
        DrawerLayout drawerLayout= (DrawerLayout) findViewById(R.id.drawer_layout);
        drawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }
}

File: app/src/main/java/in/rahulkhatri/vegcartpro/activity/HomeActivity.kt
package `in`.rahulkhatri.vegcartpro.activity

import `in`.rahulkhatri.vegcartpro.model.ItemModel
import `in`.rahulkhatri.vegcartpro.adapter.HomeAdapter
import `in`.rahulkhatri.vegcartpro.model.UserDetailModel
import `in`.rahulkhatri.vegcartpro.utils.Constants.USER_DETAIL
import `in`.rahulkhatri.vegcartpro.utils.SharedPreferenceUtils
import `in`.rahulkhatri.vegcartpro.utils.Utility.getDeviceId
import `in`.rahulkhatri.vegcartpro.utils.Utility.getDeviceManufacturer
import `in`.rahulkhatri.vegcartpro.utils.Utility.getDeviceModel
import `in`.rahulkhatri.vegcartpro.utils.Utility.getOsVersionName
import `in`.rahulkhatri.vegcartpro.R
import `in`.rahulkhatri.vegcartpro.utils.Constants.REQUEST_CODE_UPDATE
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.Bundle
import android.os.Handler
import com.google.android.material.navigation.NavigationView
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.appcompat.widget.Toolbar
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import com.daimajia.slider.library.Animations.DescriptionAnimation
import com.daimajia.slider.library.SliderLayout
import com.daimajia.slider.library.SliderTypes.BaseSliderView
import com.daimajia.slider.library.SliderTypes.BaseSliderView.OnSliderClickListener
import com.daimajia.slider.library.SliderTypes.TextSliderView
import com.daimajia.slider.library.Tricks.ViewPagerEx
import com.google.android.play.core.appupdate.AppUpdateManagerFactory
import com.google.android.play.core.install.model.AppUpdateType
import com.google.android.play.core.install.model.UpdateAvailability
import com.google.firebase.database.FirebaseDatabase
import kotlinx.android.synthetic.main.activity_home.*
import java.util.*

class HomeActivity : AppCompatActivity(), OnSliderClickListener, ViewPagerEx.OnPageChangeListener, HomeAdapter.ClickListener, NavigationView.OnNavigationItemSelectedListener {
    private val itemList: MutableList<ItemModel> = mutableListOf()
    private var recyclerView: RecyclerView? = null
    var adapter: HomeAdapter? = null
    private val main: LinearLayout? = null
    var sliderLayout: SliderLayout? = null
    var HashMapForURL: HashMap<String, String>? = null
    var HashMapForLocal: HashMap<String, Int> = HashMap()
    var doubleBackToExitPressedOnce = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
        getUserDetail()
        inAppUpdate()
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        toolbar.title = "Menu"
        setSupportActionBar(toolbar)
        sliderLayout = findViewById(R.id.slider)
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        recyclerView = findViewById(R.id.recyclerView)

        if (!isOnline()) {
            alertDialog()
        }
        val toggle = ActionBarDrawerToggle(this, drawer, toolbar, R.string.open, R.string.close)
        drawer.addDrawerListener(toggle)
        toggle.syncState()
        val navigationView = findViewById<NavigationView>(R.id.nav_view)
        navigationView.setNavigationItemSelectedListener(this)
        AddImageUrlLocal()
        // Call below method if you want to stop Automatic sliding
        for (name in HashMapForLocal.keys) {
            val textSliderView = TextSliderView(this@HomeActivity)
            textSliderView.description(name).image(HashMapForLocal[name]!!).setScaleType(BaseSliderView.ScaleType.CenterCrop).setOnSliderClickListener(this)
            textSliderView.bundle(Bundle())
            textSliderView.bundle.putString("extra", name)
            sliderLayout?.addSlider(textSliderView)
        }
        sliderLayout?.setPresetTransformer(SliderLayout.Transformer.DepthPage)
        sliderLayout?.setPresetIndicator(SliderLayout.PresetIndicators.Center_Bottom)
        sliderLayout?.setCustomAnimation(DescriptionAnimation())
        sliderLayout?.setDuration(3000)
        sliderLayout?.addOnPageChangeListener(this@HomeActivity)
        // Bottom
        prepareItem()
        adapter = HomeAdapter(itemList)
        adapter?.setClickListener(this)
        adapter?.setClickListener(this)
        //RecyclerView.LayoutManager layoutManager= new LinearLayoutManager(getApplicationContext());
// Try Grid layout for Recycler view.
        recyclerView?.layoutManager = GridLayoutManager(this, 2, GridLayoutManager.VERTICAL, false)
        recyclerView?.itemAnimator = DefaultItemAnimator()
        recyclerView?.adapter = adapter
    }

    //  In-app update
    private fun inAppUpdate() {
        val updateManager = AppUpdateManagerFactory.create(this)
        updateManager.appUpdateInfo.addOnSuccessListener {
            if (it.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE) {
                updateManager.startUpdateFlowForResult(it, AppUpdateType.IMMEDIATE, this, REQUEST_CODE_UPDATE)
            }
        }
    }

    //  get user basic Detail
    private fun getUserDetail() {
        if (isOnline()) {
            val reference = FirebaseDatabase.getInstance().getReference(USER_DETAIL)
            val userDetail = UserDetailModel(getOsVersionName(), getDeviceModel(), getDeviceManufacturer())
            reference.child(getDeviceId(this)).setValue(userDetail)
        }
    }

    fun AddImagesUrlOnline() {
        HashMapForURL = HashMap()
        HashMapForURL!!["All Veg"] = "http://"
    }

    fun AddImageUrlLocal() {
        HashMapForLocal = HashMap()
        HashMapForLocal["Fruits"] = R.drawable.fruit
        HashMapForLocal["Cauliflower"] = R.drawable.cauliflower
        HashMapForLocal["Fruit and Vegi"] = R.drawable.giphy
        HashMapForLocal["Dancing Vegi"] = R.drawable.vegjig
    }

    override fun onStop() {
        sliderLayout!!.stopAutoCycle()
        super.onStop()
    }

    override fun onSliderClick(slider: BaseSliderView) {

//        Log.e("silder clicked:", " ${slider.bundle["extra"]}")
        //  Toast.makeText(this,slider.getBundle().get("extra")+ "",Toast.LENGTH_SHORT).show();
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

    override fun onPageSelected(position: Int) {
//        Log.d("Slider Demo", "Page Change:$position")
    }

    override fun onPageScrollStateChanged(state: Int) {}

    // Bottom
    private fun prepareItem() {
//        var item = Item(R.drawable.shutterstock, "Flower", "Price is 30")
//        itemList.add(item)
        itemList.add(ItemModel(R.drawable.shutterstock, "Flower", "Price is 30"))
        itemList.add(ItemModel(R.drawable.apple, "Fruit", "Price is 150"))
        itemList.add(ItemModel(R.drawable.lettuce2, "Leaves", "Price is 70"))
        itemList.add(ItemModel(R.drawable.carrot, "Root", "Price is 30"))
        itemList.add(ItemModel(R.drawable.tomato, "Salad", "Price is 20"))
    }

    override fun itemClicked(view: View?, position: Int) {

        when (position) {
            0 -> SharedPreferenceUtils(this).setCategoryItem("flower")
            1 -> SharedPreferenceUtils(this).setCategoryItem("fruit")
            2 -> SharedPreferenceUtils(this).setCategoryItem("leaves")
            3 -> SharedPreferenceUtils(this).setCategoryItem("root")
            4 -> SharedPreferenceUtils(this).setCategoryItem("salad")
            else -> Log.e("position :", position.toString())
        }
        startActivity(Intent(this@HomeActivity, CategoryItemActivity::class.java).putExtra("ItemPosition ", position))
    }

    // Navigation Drawer
    override fun onBackPressed() {
//        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        if (drawer_layout.isDrawerOpen(GravityCompat.START)) {
            drawer_layout.closeDrawer(GravityCompat.START)
        } else {
            if (doubleBackToExitPressedOnce) {
                super.onBackPressed()
                moveTaskToBack(true)
                return
            }
            doubleBackToExitPressedOnce = true
            Toast.makeText(this, "Please click BACK again to exit", Toast.LENGTH_SHORT).show()
            Handler().postDelayed({ doubleBackToExitPressedOnce = false }, 2000)
            //  super.onBackPressed();
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.fruit -> navigationItemActivity("fruit")
            R.id.salid -> navigationItemActivity("salad")
            R.id.flower -> navigationItemActivity("flower")
            R.id.leaves -> navigationItemActivity("leaves")
            R.id.root -> navigationItemActivity("root")
            R.id.nav_send -> startActivity(Intent(this@HomeActivity, ContactActivity::class.java))
            R.id.mycart -> startActivity(Intent(this@HomeActivity, MyCartActivity::class.java))
        }
        val drawerLayout = findViewById<androidx.drawerlayout.widget.DrawerLayout>(R.id.drawer_layout)
        drawerLayout.closeDrawer(GravityCompat.START)
        return true
    }

    private fun navigationItemActivity(stSharedPref: String) {
        SharedPreferenceUtils(this).setCategoryItem(stSharedPref)
        startActivity(Intent(this@HomeActivity, CategoryItemActivity::class.java))
    }

    private fun alertDialog() {
        val builder = AlertDialog.Builder(this@HomeActivity)
        builder.setMessage("Please Connect to Internet")
        builder.setCancelable(false).setIcon(R.drawable.ic_internet).setTitle("No Internet Connection")
        builder.setPositiveButton("OK"
        ) { dialog, id ->
            dialog.cancel()
            finish()
        }.setNegativeButton("Retry") { dialog, which ->
            if (!isOnline()) {
                alertDialog()
            } else {
                dialog.dismiss()
            }
        }
        val alert = builder.create()
        alert.show()
    }

    private fun isOnline(): Boolean {
        val result: Boolean
        val connectivityManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val networkCapabilities = connectivityManager.activeNetwork ?: return false
            val activeNetwork = connectivityManager.getNetworkCapabilities(networkCapabilities)
                    ?: return false

            result = when {
                activeNetwork.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> true
                activeNetwork.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> true
                activeNetwork.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> true
                else -> false
            }
        } else {
            result = connectivityManager.activeNetworkInfo != null && connectivityManager.activeNetworkInfo!!.isConnected
//            connectivityManager.run {
//                connectivityManager.activeNetworkInfo ?.run {
//                    result = when(type) {
//                        ConnectivityManager.TYPE_WIFI -> true
//                        ConnectivityManager.TYPE_MOBILE -> true
//                        ConnectivityManager.TYPE_ETHERNET -> true
//                        else -> false
//                    }
//                }
//            }
        }
        return result
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.home_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.bucket -> {
                startActivity(Intent(this, CartActivity::class.java))
                true
            }
            else -> super.onOptionsItemSelected(item)
        }

    }
}